### PR TITLE
BaseModel, FancyFormsModule, Modal, Button

### PR DESCRIPTION
**Pull request by sandikbarr**
BaseModel, FancyForms, and Modal #39 and #34 are interdependent. Decided I may as well add #26 buttons to the FancyFormsModule to round it out while I'm at it.

### CODE REVIEW

File: src/demo/app/app.routing.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@ import { Routes, RouterModule } from '@angular/router';
 
 import { TocComponent } from './toc.component';
 import { AdvancedConfirmDemoComponent } from './fancy-form/advanced-confirm-demo.component';
+import { ButtonDemoComponent } from './fancy-form/button-demo.component';
```
Comment (gcampo88) [inline src/demo/app/app.routing.ts]: dup import name on line 21 preventing `npm start`! 
I assume that we should be using `fancy-form/button-demo.component`  (and should delete `global-css/button-demo.component` and its import)?

Comment (sandikbarr) [inline src/demo/app/app.routing.ts]: doh! these are two different demos. Will rename!

File: src/demo/app/fancy-form/advanced-confirm-demo.component.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      <aside>
+        Usage:
+        <pre class="code">
+          &lt;input
+              prxAdvancedConfirm="That's a risky move" 
+              [prxModel]="model" 
+              prxName="foo" (change)="model.set('foo', $event)"&gt;
+        </pre>
+        Example:
+        <input prxAdvancedConfirm="That's a risky move" [prxModel]="model" prxName="foo" (change)="model.set('foo', $event)">
```
Comment (gcampo88) [inline src/demo/app/fancy-form/advanced-confirm-demo.component.ts]: should this input be showing 'bar' before it's changed? (and - should pressing cancel discard the changes?)

Comment (sandikbarr) [inline src/demo/app/fancy-form/advanced-confirm-demo.component.ts]: It will if I hook it up with ngModel. Done!

Comment (gcampo88) [inline src/demo/app/fancy-form/advanced-confirm-demo.component.ts]: 👍

File: src/demo/app/fancy-form/fancy-field-demo.component.ts [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        <li>
+          <code>@Input() hideinvalid: boolean</code>
+           - set to false if validation messages are not desired
+        </li>
+        <li>
+          <code>@Input() advancedConfirm: string</code>
+           - a confirmation message for "advanced" fields
+        </li>
+        <li>
+          <code>@Input() strict: boolean</code>
```
Comment (gcampo88) [inline src/demo/app/fancy-form/fancy-field-demo.component.ts]: could be worth explaining meaning of strict? (strict:true will prevent save if invalid, right?)

File: src/demo/app/fancy-form/fancy-field-demo.component.ts [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+      </aside>
+      
+      <aside>
+        <h2><code>textarea</code></h2>
+        <p>Use textarea for longer text fields.</p>
+        Usage:
+        <pre class="code">
+          &lt;prx-fancy-field textarea [model]="model" name="script" label="Your script"&gt;
+            &lt;div class="fancy-hint"&gt;
+              This sponsor does not require final script approval.
```
Comment (gcampo88) [inline src/demo/app/fancy-form/fancy-field-demo.component.ts]: Are we in the future?! :)

File: src/demo/app/fancy-form/fancy-field-demo.component.ts [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+          <div class="fancy-hint">
+            This sponsor does not require final script approval.
+            Save your script here for storage and convenience.
+          </div>
+        </prx-fancy-field>
+      </aside>
+      
+      <aside>
+        <h2><code>.nested</code></h2>
+        <p>Use content projection with the <code>.nested</code> selector to include anything in a fancy-field. Maybe a date picker?</p>
```
Comment (gcampo88) [inline src/demo/app/fancy-form/fancy-field-demo.component.ts]: are we doing anything with the .nested selector here in this example? or maybe i'm not 150% clear on how it should be used...

File: src/demo/app/fancy-form/fancy-field-demo.component.ts [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+      </aside>
+
+      <aside>
+        <h2>no <code>model</code></h2>
+        <p>If model is undefined or not provided, the field will be disabled.</p>
+        Usage:
+        <pre class="code">
+          &lt;prx-fancy-field textinput name="nada" label="Nope"&gt;
+            &lt;div class="fancy-hint"&gt;This field is disabled because there is no model.&lt;/div&gt;
+          &lt;/prx-fancy-field&lt;
```
Comment (gcampo88) [inline src/demo/app/fancy-form/fancy-field-demo.component.ts]: closing bracket points the wrong way 
image

File: src/demo/app/fancy-form/fancy-field-demo.component.ts [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+      
+      <aside>
+        <h2><code>.fancy-hint</code></h2>
+        <p>Use content projection with the <code>.fancy-hint</code> selector to show a field hint. Hints typically end with punctuation.</p>
+        Usage:
+        <pre class="code">
+          &lt;div class="fancy-hint"&gt;Hint text&lt;/div&gt;
+        </pre>
+        Example:
+        <prx-fancy-field textinput name="showhint" label="Hint">
```
Comment (gcampo88) [inline src/demo/app/fancy-form/fancy-field-demo.component.ts]: I think in the other examples, the Example matches the code shown in the Usage -- not here?

File: src/demo/app/fancy-form/fancy-field-demo.component.ts [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+})
+
+export class FancyFieldDemoComponent {
+  model = new DemoModel(null,
+    new HalDoc({
+      foo: 'bar',
+      complete: true,
+      count: 11,
+      color: 'yellow',
+      script: 'Visit blue apron dot com slash exciting show name to get your first three deliveries free.',
```
Comment (gcampo88) [inline src/demo/app/fancy-form/fancy-field-demo.component.ts]: :)

File: src/demo/app/fancy-form/fancy-field-demo.component.ts [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+      arrival: new Date(),
+      feedId: '196',
```
Comment (gcampo88) [inline src/demo/app/fancy-form/fancy-field-demo.component.ts]: feedId? is this used?

File: src/demo/app/fancy-form/fancy-field-demo.component.ts (hunk 4 above)
Comment (sandikbarr) [inline src/demo/app/fancy-form/fancy-field-demo.component.ts]: You didn't know about my tardis?

File: src/demo/app/fancy-form/fancy-field-demo.component.ts (hunk 9 above)
Comment (sandikbarr) [inline src/demo/app/fancy-form/fancy-field-demo.component.ts]: This is a really good catch! I knew you would do a thorough job of looking at these things I had become blind to. I was actually using this feedId at one point to demo `.prefix` and `.suffix` selectors but they didn't look quite right, which made me remember that I had created those selectors months ago for setting just the feedId portion of the URL, but then we didn't end up using that so those selectors were mostly removed from fancy-field and had no styling. They were still in the template but have been removed completely now.

File: src/demo/app/model/base-model-demo.component.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+            <code>decode(): void</code> is called from the init() method to set the model fields from the HalDoc fields
+          </li>
+          <li>
+            <code>encode(): &#123;&#125;</code> returns an object with the model's data used in persisting the underlying HalDoc
+          </li>
+          <li>
+            <code>saveNew(data: &#123;&#125;): Observable&lt;HalDoc&gt;</code> should call the HalDoc create method
+          </li>
+        </ul>
+      </section>
```
Comment (gcampo88) [inline src/demo/app/model/base-model-demo.component.ts]: Could throw in an Example with a link to or screenshot of `src/demo/app/model/demo.model.ts` if you want?

File: src/demo/app/fancy-form/fancy-field-demo.component.ts (hunk 5 above)
Comment (sandikbarr) [inline src/demo/app/fancy-form/fancy-field-demo.component.ts]: This is an incorrect explanation. Fixed!

File: src/demo/app/spinner/spinner-demo.component.ts [hunk 11]
```diff
@@ -8,13 +8,20 @@ import {Component} from '@angular/core';
...
         The animation uses <code>position: absolute</code>, so it should be inside a <code>position: relative</code> container. 
       </p>
+      <dl>
+        <dt>module</dt><dd><code>SpinnerModule</code></dd>
+        <dt>selector</dt><dd><code>prx-spinner</code></dd>
+      </dl>
+      <ul>
+        <li><code>@Input() set spinning(value: boolean)</code> - toggles the animation</li>
+        <li><code>@Input() delay: number</code> - defaults to 300 seconds</li>
+        <li><code>@Input() inverse: boolean</code> - defaults to false but when true will show a grey background color</li>
```
Comment (gcampo88) [inline src/demo/app/spinner/spinner-demo.component.ts]: looks like `inverse=true` also has some extra margin? 
image

File: src/demo/app/model/demo.model.ts [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+import { BaseModel, RelatedMap, HalDoc, REQUIRED } from 'ngx-prx-styleguide';
+
+export class DemoModel extends BaseModel {
+  public foo: string;
+  public complete: boolean;
+  public count: number;
+  public color: string;
+  public script: string;
+  public arrival: Date;
+  public feedId: string;
```
Comment (gcampo88) [inline src/demo/app/model/demo.model.ts]: `feedId` still in demoModel setable + decode but no longer used in fancy field -- maybe for use later?

File: src/demo/app/spinner/spinner-demo.component.ts (hunk 11 above)
Comment (sandikbarr) [inline src/demo/app/spinner/spinner-demo.component.ts]: It does. I can't really explain why it's done this way for the prx header. Ryan had actually asked this same question. This wasn't as much an effort to undo our past mistakes as to give us the ability to move forward baggage and all.

File: src/demo/app/model/base-model-demo.component.ts (hunk 10 above)
Comment (sandikbarr) [inline src/demo/app/model/base-model-demo.component.ts]: DemoModel just isn't a very good example. I agree that the BaseModel documentation is not really complete, but I also feel like I'm getting bogged down in documentation. I'll address these trade offs in B/W, and we'll make a decision about how much more work should be done here right now.

Comment (sandikbarr) [inline src/demo/app/model/base-model-demo.component.ts]: Calling this good enough for now, but will address later with #45. I do think this is important, especially for new developers on this project, but maybe not as valuable right now for those of us who have gotten to know it a little better. We'll aim to provide solid documentation on components but less so on the existing non angular stuff in this first round of getting the styleguide off the ground.

**Comment by sandikbarr** [comment]
@gcampo88 I dunno on that spinner thing, pre-existing condition. feedId is all the way gone now, sorry for that late Friday half assed cleanup. We're gonna call BaseModel docs good enough for now, opened a ticket to address at a later date:  #45 So back to you!

File: src/demo/app/model/base-model-demo.component.ts (hunk 10 above)
Comment (gcampo88) [inline src/demo/app/model/base-model-demo.component.ts]: 👍

File: src/demo/app/spinner/spinner-demo.component.ts (hunk 11 above)
Comment (gcampo88) [inline src/demo/app/spinner/spinner-demo.component.ts]: good call

**Comment by gcampo88** [comment]
💯

**Pull request merged**